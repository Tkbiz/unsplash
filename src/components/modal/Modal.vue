<template>
  <v-layout
      row
      justify-center>
    <v-dialog
        content-class="custom-dialog"
        persistent
        v-model="showModal">
      <v-btn
          class="btn-close"
          icon
          @click.prevent="closeModal">
        <v-icon>
          mdi-close
        </v-icon>
      </v-btn>
      <div class="dialog-body">
        <v-col class="image-block">
          <v-img
              v-if="card.urls"
              class="main-img"
              :src="card.urls.regular ? card.urls.regular : '@/assets/images/no-image.png'">
          </v-img>
          <v-col class="action-block">
            <v-btn
                class="mr-2 btn-like"
                :class="{'btn-active': isLiked}"
                icon
                @click.prevent="addLike(card.id)">
              <v-icon>mdi-heart</v-icon>
            </v-btn>
            <v-btn
                class="btn-add"
                :class="{'btn-active': isAddedToCollection}"
                icon
                @click.prevent="openCollectionsList">
              <v-icon>mdi-plus</v-icon>
            </v-btn>
            <v-alert
                class="custom-alert"
                v-if="showMessage"
                type="warning">
              Please log in to do it!
            </v-alert>
            <v-alert
                class="collections-alert"
                v-model="showCollectionsList"
                text>
              <v-card>
                <v-card-title
                    v-for="collection in userCollections"
                    :key="collection.id"
                    @click.prevent="addToCollection(collection.id, card.id)"
                    class="headline cursor-pointer">
                  {{ collection.title }}
                </v-card-title>
              </v-card>
            </v-alert>
          </v-col>
        </v-col>
        <v-col v-if="card.description"
               cols="12"
               class="py-0 stats-block">
          <v-row>
            <div class="col-12 stats-info stats-description">
              {{ card.description }}
            </div>
          </v-row>
        </v-col>
        <UserModalInfo
            :userCard="card.user"/>
        <ModalStats
            :photoStats="photoStats"
            :tags="card.tags"/>
        <v-col cols="12 mb-6">
          <v-row>
            <div v-if="photoStats.views">
              <ViewsChart
                  :photoViews="photoStats.views"/>
            </div>
            <div v-if="photoStats.likes">
              <LikesChart
                  :photoLikes="photoStats.likes"/>
            </div>
          </v-row>
        </v-col>
        <v-col v-if="samePhotos"
               cols="12"
               class="pa-0">
          <v-row justify="start">
            <div class="block-img"
                 v-for="item in samePhotos"
                 :key="item.id">
              <img
                  class="image"
                  :src="item.urls.thumb ? item.urls.thumb : '@/assets/images/no-image.png'"
                  alt="card image">
            </div>
          </v-row>
        </v-col>
        <Loader v-if="loading"/>
        <v-col v-if="sameNewPhotos"
               cols="12"
               class="py-0">
          <v-row
              justify="center">
            <v-btn
                icon
                @click.prevent="showMoreImages">
              <v-icon
                  class="dots">
                mdi-dots-horizontal
              </v-icon>
            </v-btn>
          </v-row>
        </v-col>
      </div>
    </v-dialog>
  </v-layout>
</template>

<script>
import ViewsChart from '@/components/charts/ViewsChart'
import LikesChart from '@/components/charts/LikesChart'
import ModalStats from '@/components/modal/ModalStats'
import UserModalInfo from '@/components/modal/UserModalInfo'

export default {
  name: 'Modal',
  props: {
    visibleModal: {
      type: Boolean,
      required: true
    },
    card: {
      type: Object,
      required: true
    },
    samePhotos: {
      type: Array,
      required: true
    },
    tagName: {
      type: String,
      required: true
    },
    photoStats: {
      type: Object,
      required: true
    },
    userName: {
      type: String
    }
  },
  components: {
    UserModalInfo,
    ModalStats,
    ViewsChart,
    LikesChart
  },
  data: () => ({
    page: 1,
    sameNewPhotos: [],
    loading: false,
    showMessage: false,
    showCollectionsList: false
  }),
  methods: {
    closeModal() {
      this.showModal = false
    },
    async showMoreImages() {
      this.loading = true
      this.page++
      this.sameNewPhotos = await this.$store.dispatch('searchPhotos', {
        tagName: this.tagName,
        page: this.page
      })
      this.sameNewPhotos.map(newPhoto => this.samePhotos.push(newPhoto))
      this.loading = false
    },
    async addLike(id) {
      if (this.isLoggedIn) {
        if (this.isLiked !== true) {
          await this.$store.dispatch('addLike', id)
        }
      } else {
        this.showMessage = true
        setTimeout(() => {
          this.showMessage = false
        }, 2000)
      }
    },
    async openCollectionsList() {
      if (this.isLoggedIn) {
        this.showCollectionsList = true
        await this.$store.dispatch('fetchUserCollections', this.userName)
      } else {
        this.showMessage = true
        setTimeout(() => {
          this.showMessage = false
        }, 2000)
      }
    },
    async addToCollection(collectionId, photoId) {
      await this.$store.dispatch('addToCollection', {collectionId, photoId})
      this.showCollectionsList = false
    }
  },
  computed: {
    showModal: {
      get() {
        return this.visibleModal
      },
      set(value) {
        if (!value) {
          this.$emit('close')
        }
      }
    },
    isLoggedIn() {
      return this.$store.getters.accessToken
    },
    userCollections() {
      return this.$store.getters.userCollections
    },
    isAddedToCollection() {
      return this.$store.getters.isAddedToCollection
    },
    isLiked() {
      return this.$store.getters.isLiked
    }
  }
}
</script>
