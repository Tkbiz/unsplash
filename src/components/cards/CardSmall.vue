<template>
  <v-col
      v-if="card"
      class="custom-card pa-0">
    <img @click.prevent="openModal(card)"
         class="card-img"
         :src="card.urls.small ? card.urls.small : '@/assets/images/no-image.png'"
         alt="card image">
    <div @click.prevent="openModal(card)"
         class="card-body">
      <div v-if="card.user.name"
           :class="{'mb-6': !card.user.location && !card.description}"
           class="card-title">
        {{ card.user.name }}
      </div>
      <div v-if="card.user.location"
           :class="{'mb-6': !card.description}"
           class="card-subtitle">
        {{ card.user.location }}
      </div>
      <div v-if="card.description"
           class="card-text">
        {{card.description | truncate(140, '...')}}
      </div>
    </div>
    <div class="card-footer">
      <div class="ml-n2">
        <v-btn
            @click.prevent="downloadPhoto(card.urls.small)"
            icon>
          <v-icon>
            mdi-bookmark-outline
          </v-icon>
        </v-btn>
      </div>
      <div>
        <div>
          <span class="mr-5">
            <v-btn
                @click.prevent="setLike"
                icon>
              <v-icon>
                mdi-heart-outline
              </v-icon>
            </v-btn>
            <span class="numbers">
              {{ card.likes }}
            </span>
          </span>
          <span>
            <v-btn
                @click.prevent="setComment"
                icon>
              <v-icon>
                mdi-message-outline
              </v-icon>
            </v-btn>
            <span class="numbers">
              {{ card.downloads }}
            </span>
          </span>
        </div>
      </div>
    </div>
    <Modal
        v-if="modal"
        :photoStats="photoStats"
        :tagName="tagName"
        :samePhotos="samePhotos"
        :card="card"
        :userName="userName"
        :visibleModal="modal"
        @close="closeModal"/>
  </v-col>
</template>

<script>
import Modal from '@/components/modal/Modal'
import {truncateTextMixin} from '@/mixins/truncateTextMixin'

export default {
  name: 'CardSmall',
  mixins: [truncateTextMixin],
  props: {
    card: {
      type: Object,
      required: true
    },
    userName: {
      required: true
    }
  },
  components: {
    Modal
  },
  data: () => ({
    modal: false,
    tagName: '',
    samePhotos: [],
    photoStats: {},
    page: 1
  }),
  methods: {
    async openModal() {
      this.modal = true
      this.card.tags.forEach(tag => this.tagName = tag.title)
      this.samePhotos = await this.$store.dispatch('searchPhotos', {
        tagName: this.tagName,
        page: this.page
      })
      this.photoStats = await this.$store.dispatch('fetchPhotoStats', this.card.id)
    },
    closeModal() {
      this.modal = false
    },
    setLike() {
      console.log('was liked!')
    },
    setComment() {
      console.log('comment was written!')
    },
    async downloadPhoto(url) {
      await this.$store.dispatch('downloadPhoto', url)
    }
  }
}
</script>

<style scoped>

</style>
