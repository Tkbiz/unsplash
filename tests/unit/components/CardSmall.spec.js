import {mount, createLocalVue} from '@vue/test-utils'
import Vue from 'vue'
import Vuex from 'vuex'
import Vuetify from 'vuetify'
import CardSmall from '@/components/cards/CardSmall'
import Modal from '@/components/modal/Modal'
import {truncateTextMixin} from '@/mixins/truncateTextMixin'

Vue.use(Vuetify)

const localVue = createLocalVue()
localVue.use(Vuex)

describe('CardSmall.vue', () => {
  let vuetify
  let wrapper
  let actions
  let store

  beforeEach(() => {
    vuetify = new Vuetify()

    actions = {
      searchPhotos: jest.fn(),
      fetchPhotoStats: jest.fn(),
      downloadPhoto: jest.fn()
    }
    store = new Vuex.Store({
      actions
    })

    const el = document.createElement('div')
    el.setAttribute('data-app', true)
    document.body.appendChild(el)

    wrapper = mount(CardSmall, {
      localVue,
      vuetify,
      propsData: {
        card: {
          urls: {},
          user: {},
          tags: []
        },
        userName: ''
      },
      data: () => ({
        modal: false,
        tagName: '',
        samePhotos: [],
        photoStats: {},
        page: 1
      }),
      store
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy()
  })

  it("renders truncateTextMixin", () => {
    expect(wrapper.exists(truncateTextMixin)).toBe(true)
  })

  it("renders Modal", () => {
    expect(wrapper.exists(Modal)).toBe(true)
  })

  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders props when passed', () => {
    const card = {
      urls: {},
      user: {},
      tags: []
    }
    const userName = ''
    expect(wrapper.props()).toEqual({card, userName})
  })

  // it('dispatches a searchPhotos action when image is clicked', async () => {
  //   const tagName = ''
  //   const page = 1
  //
  //   wrapper.find('.card-img').trigger('click')
  //   await wrapper.vm.$nextTick()
  //
  //   expect(actions.searchPhotos.mock.calls).toHaveLength(1)
  //   expect(actions.searchPhotos.mock.calls[0][1]).toEqual({tagName, page})
  //   expect(actions.searchPhotos).toHaveBeenCalled()
  // })
  //
  // it('dispatches a searchPhotos action when card is clicked', async () => {
  //   const tagName = ''
  //   const page = 1
  //
  //   wrapper.find('.card-body').trigger('click')
  //   await wrapper.vm.$nextTick()
  //
  //   expect(actions.searchPhotos.mock.calls).toHaveLength(1)
  //   expect(actions.searchPhotos.mock.calls[0][1]).toEqual({tagName, page})
  //   expect(actions.searchPhotos).toHaveBeenCalled()
  // })
  //
  // it('dispatches a fetchPhotoStats action when image is clicked', async () => {
  //   wrapper.find('.card-img').trigger('click')
  //   await wrapper.vm.$nextTick()
  //
  //   expect(actions.fetchPhotoStats.mock.calls).toHaveLength(1)
  //   expect(actions.fetchPhotoStats).toHaveBeenCalled()
  // })
  //
  // it('dispatches a fetchPhotoStats action when card is clicked', async () => {
  //   wrapper.find('.card-body').trigger('click')
  //   await wrapper.vm.$nextTick()
  //
  //   expect(actions.fetchPhotoStats.mock.calls).toHaveLength(1)
  //   expect(actions.fetchPhotoStats).toHaveBeenCalled()
  // })

  it('dispatches a downloadPhoto action when button is clicked', async () => {
    wrapper.find('.card-footer .ml-n2 .v-btn').trigger('click')
    await wrapper.vm.$nextTick()

    expect(actions.downloadPhoto.mock.calls).toHaveLength(1)
    expect(actions.downloadPhoto).toHaveBeenCalled()
  })
})
