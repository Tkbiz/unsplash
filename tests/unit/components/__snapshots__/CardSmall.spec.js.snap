// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardSmall.vue should match snapshot 1`] = `
<div class="custom-card pa-0 col"><img src="@/assets/images/no-image.png" alt="card image" class="card-img">
  <div class="card-body">
    <!---->
    <!---->
    <!---->
  </div>
  <div class="card-footer">
    <div class="ml-n2"><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-bookmark-outline theme--light"></i></span></button></div>
    <div>
      <div><span class="mr-5"><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-heart-outline theme--light"></i></span></button> <span class="numbers">

          </span></span> <span><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-message-outline theme--light"></i></span></button> <span class="numbers">

          </span></span></div>
    </div>
  </div>
  <!---->
</div>
`;
